ports/esp8266/esp_espnow.c

```diff
@@ -0,0 +1,255 @@
+ */
+
+
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "c_types.h"
+#include "espnow.h"
+
+#include "py/runtime.h"
+#include "py/mphal.h"
+#include "py/nlr.h"
+#include "py/objlist.h"
+#include "py/runtime.h"
+#include "py/mphal.h"
+#include "py/mperrno.h"
+
+#define ESP_NOW_ETH_ALEN             6         /*!< Length of ESPNOW peer MAC address */
+#define ESP_NOW_KEY_LEN              16        /*!< Length of ESPNOW peer local master key */
+#define ESP_NOW_MAX_DATA_LEN         250       /*!< Maximum length of ESPNOW data which is sent very time */
+
+enum mt_tx_status {
+    MT_TX_STATUS_OK = 0,
+    MT_TX_STATUS_FAILED,
+};
+
+static inline void esp_espnow_exceptions(int e) {
+    if (e != 0) {
+        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_OSError, "ESP-Now Unknown Error 0x%04x", e));
+    }
+}
+
+#define ESPNOW_EXCEPTIONS(x) do { esp_espnow_exceptions(x); } while (0);
+
+#define ESP_NOW_PEER_EXIST(addr_buf) do {                                                           \
+    int e = esp_now_is_peer_exist(addr_buf);                                                        \
+    if (e == 0) {                                                                                   \
+        mp_raise_msg(&mp_type_OSError, "ESP-Now Peer Not Found");                                   \
+    } else if (e < 0) {                                                                             \
+        esp_espnow_exceptions(e);                                                                   \
+    }                                                                                               \
+} while (0);                                                                                        \
+
+static inline void _get_bytes(mp_obj_t str, size_t len, uint8_t *dst) {
+    size_t str_len;
+    const char *data = mp_obj_str_get_data(str, &str_len);  
+    if (str_len != len) mp_raise_ValueError("bad len");
+    memcpy(dst, data, len);
+}
+
+static mp_obj_t send_cb_obj = mp_const_none;
+static mp_obj_t recv_cb_obj = mp_const_none;
+
+STATIC void send_cb(u8 *mac_addr, u8 status)
+{
+    if (send_cb_obj != mp_const_none) {
+        mp_obj_tuple_t *msg = mp_obj_new_tuple(2, NULL);
+        msg->items[0] = mp_obj_new_bytes(mac_addr, ESP_NOW_ETH_ALEN);
+        msg->items[1] = (status == MT_TX_STATUS_OK) ? mp_const_true : mp_const_false;
+        mp_sched_schedule(send_cb_obj, msg);
+    }
+}
+
+STATIC void recv_cb(u8 *mac_addr, u8 *data, u8 len)
+{
+    if (recv_cb_obj != mp_const_none) {
+        mp_obj_tuple_t *msg = mp_obj_new_tuple(2, NULL);
+        msg->items[0] = mp_obj_new_bytes(mac_addr, ESP_NOW_ETH_ALEN);
+        msg->items[1] = mp_obj_new_bytes(data, len);
+        mp_sched_schedule(recv_cb_obj, msg);
+    }
+} 
+
+static int initialized = 0;
+
+STATIC mp_obj_t espnow_init() {
+    if (!initialized) {
+        ESPNOW_EXCEPTIONS(esp_now_init());
+        ESPNOW_EXCEPTIONS(esp_now_set_self_role(ESP_NOW_ROLE_COMBO));
```

Comment:
does it have any disadvantages to use the COMBO mode if I only want to send (or receive)?

Reply:
This option is available only in esp8266 (always COMBO mode in esp32).
Just trying to be a bit more uniform :)